Importing a file outside of the project folder causes: Emit skipped error

Im trying to import a file outside of the project. It's a shared file between the frontend and the server.

When i try import it it throws this error:

```
node_modules/ts-node/src/index.ts:370
        throw new TypeError(`${relative(cwd, fileName)}: Emit skipped`)
              ^
TypeError: project/my_file.ts: Emit skipped
    at getOutput (/Users/.../node_modules/ts-node/src/index.ts:370:15)
    at Object.compile (/Users/.../node_modules/ts-node/src/index.ts:558:11)
    at Module.m._compile (/Users/.../node_modules/ts-node/src/index.ts:439:43)
    at Module._extensions..js (internal/modules/cjs/loader.js:713:10)
    at Object.require.extensions.(anonymous function) [as .ts] (/Users/.../node_modules/ts-node/src/index.ts:442:12)
    at Module.load (internal/modules/cjs/loader.js:612:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:551:12)
    at Function.Module._load (internal/modules/cjs/loader.js:543:3)
    at Module.require (internal/modules/cjs/loader.js:650:17)
    at require (internal/modules/cjs/helpers.js:20:18)
```

I'm able to import it from my angular project which is a sibling.

I've tried adding the file to my tsconfig under includes and files. This is what my tsconfig looks like:

```JavaScript
{
  "extends": "../tsconfig.json",
  "compilerOptions": {
    "outDir": "../../dist/backend",
    "rootDir": "./",
    "module": "commonjs",
    "types": [
      "node"
    ]
  },
  "files": [
    "../lib/shared_file_to_import.ts" // Can't import this file
  ],
  "include": [
    "./**/*.ts"
  ]
}
```

And this is my tsconfig file that the project extends:

```JavaScript
{
  "compileOnSave": false,
  "extends": "./node_modules/gts/tsconfig-google.json",
  "compilerOptions": {
    "baseUrl": "./",
    "rootDir": ".",
    "outDir": "./dist",
    "sourceMap": true,
    "strict": true,
    "strictPropertyInitialization": false,
    "declaration": false,
    "moduleResolution": "node",
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "noImplicitAny": true,
    "target": "es5",
    "strictNullChecks": true,
    "typeRoots": [
      "node_modules/@types"
    ],
    "lib": [
      "es2017",
      "dom"
    ]
  },
  "include": [
    "./lib/shared_file_to_import.ts"
  ]
}
```

I have a similar problem while trying to import a `.json` file without `resolveJsonModule: true` in `tscofig.json`. I'm using [this setup](https://stackoverflow.com/questions/55753163/package-json-is-not-under-rootdir/61426303#61426303) to be able to import JSON files from outside he rootDir. It works fine with `tsc`, and if I add `resolveJsonModule: true` in `tsconfig.json`, it fails with '../lib/settings.json' is not under 'rootDir'. So `resolveJsonModule: true` must *not* be in `tsconfig.json`.

When I run `ts-node` on a file that depends on `settings.json`, I get this error from ts-node:

> TSError: ⨯ Unable to compile TypeScript:
src/modules/MyModule.ts(21,27): error TS2732: Cannot find module '../../../lib/settings.json'. Consider using '--resolveJsonModule' to import module with '.json' extension

--- a/.prettierignore
+++ b/.prettierignore
@@ -11,3 +11,4 @@ tests/throw error.ts
 tests/throw error react tsx.tsx
 tests/esm/throw error.ts
 tests/legacy-source-map-support-interop/index.ts
+tests/main-realpath/symlink/symlink.tsx

--- a/src/index.ts
+++ b/src/index.ts
@@ -1016,7 +1016,7 @@ export function create(rawOptions: CreateOptions = {}): Service {
         if (diagnosticList.length) reportTSError(diagnosticList);
 
         if (output.emitSkipped) {
-          throw new TypeError(`${relative(cwd, fileName)}: Emit skipped`);
+          return [undefined, undefined, true];
         }
 
         // Throw an error when requiring `.d.ts` files.
@@ -1029,7 +1029,7 @@ export function create(rawOptions: CreateOptions = {}): Service {
           );
         }
 
-        return [output.outputFiles[1].text, output.outputFiles[0].text];
+        return [output.outputFiles[1].text, output.outputFiles[0].text, false];
       };
 
       getTypeInfo = (code: string, fileName: string, position: number) => {
@@ -1159,7 +1159,8 @@ export function create(rawOptions: CreateOptions = {}): Service {
       };
 
       getOutput = (code: string, fileName: string) => {
-        const output: [string, string] = ['', ''];
+        let outText = '';
+        let outMap = '';
 
         updateMemoryCache(code, fileName);
 
@@ -1179,9 +1180,9 @@ export function create(rawOptions: CreateOptions = {}): Service {
           sourceFile,
           (path, file, writeByteOrderMark) => {
             if (path.endsWith('.map')) {
-              output[1] = file;
+              outMap = file;
             } else {
-              output[0] = file;
+              outText = file;
             }
 
             if (options.emit) sys.writeFile(path, file, writeByteOrderMark);
@@ -1192,11 +1193,11 @@ export function create(rawOptions: CreateOptions = {}): Service {
         );
 
         if (result.emitSkipped) {
-          throw new TypeError(`${relative(cwd, fileName)}: Emit skipped`);
+          return [undefined, undefined, true];
         }
 
         // Throw an error when requiring files that cannot be compiled.
-        if (output[0] === '') {
+        if (outText === '') {
           if (program.isSourceFileFromExternalLibrary(sourceFile)) {
             throw new TypeError(
               `Unable to compile file from external library: ${relative(
@@ -1214,7 +1215,7 @@ export function create(rawOptions: CreateOptions = {}): Service {
           );
         }
 
-        return output;
+        return [outText, outMap, false];
       };
 
       getTypeInfo = (code: string, fileName: string, position: number) => {
@@ -1291,7 +1292,7 @@ export function create(rawOptions: CreateOptions = {}): Service {
       );
       if (diagnosticList.length) reportTSError(diagnosticList);
 
-      return [result.outputText, result.sourceMapText as string];
+      return [result.outputText, result.sourceMapText as string, false];
     };
   }
 
@@ -1308,6 +1309,7 @@ export function create(rawOptions: CreateOptions = {}): Service {
       : createTranspileOnlyGetOutputFunction(
           ts.ModuleKind.ES2020 || ts.ModuleKind.ES2015
         );
+  const getOutputTranspileOnly = createTranspileOnlyGetOutputFunction();
 
   // Create a simple TypeScript compiler proxy.
   function compile(code: string, fileName: string, lineOffset = 0) {
@@ -1315,17 +1317,19 @@ export function create(rawOptions: CreateOptions = {}): Service {
     const classification =
       moduleTypeClassifier.classifyModule(normalizedFileName);
     // Must always call normal getOutput to throw typechecking errors
-    let [value, sourceMap] = getOutput(code, normalizedFileName);
+    let [value, sourceMap, emitSkipped] = getOutput(code, normalizedFileName);
     // If module classification contradicts the above, call the relevant transpiler
     if (classification.moduleType === 'cjs' && getOutputForceCommonJS) {
       [value, sourceMap] = getOutputForceCommonJS(code, normalizedFileName);
     } else if (classification.moduleType === 'esm' && getOutputForceESM) {
       [value, sourceMap] = getOutputForceESM(code, normalizedFileName);
+    } else if (emitSkipped) {
+      [value, sourceMap] = getOutputTranspileOnly(code, normalizedFileName);
     }
     const output = updateOutput(
-      value,
+      value!,
       normalizedFileName,
-      sourceMap,
+      sourceMap!,
       getExtension
     );
     outputCache.set(normalizedFileName, { content: output });
@@ -1456,7 +1460,7 @@ function registerExtension(
 /**
  * Internal source output.
  */
-type SourceOutput = [string, string];
+type SourceOutput = [string, string, false] | [undefined, undefined, true];
 
 /**
  * Update the output remapping the source map.

--- a/src/test/index.spec.ts
+++ b/src/test/index.spec.ts
@@ -1259,7 +1259,7 @@ test.suite('ts-node', (test) => {
 
 test('Falls back to transpileOnly when ts compiler returns emitSkipped', async () => {
   const { err, stdout } = await exec(
-    `${CMD_TS_NODE_WITHOUT_PROJECT_FLAG} ./outside-rootDir/foo.js`,
+    `${CMD_TS_NODE_WITHOUT_PROJECT_FLAG} --project tsconfig.json ./outside-rootDir/foo.js`,
     {
       cwd: join(TEST_DIR, 'emit-skipped-fallback'),
     }

--- a/src/test/index.spec.ts
+++ b/src/test/index.spec.ts
@@ -1256,3 +1256,14 @@ test.suite('ts-node', (test) => {
     }
   });
 });
+
+test('Falls back to transpileOnly when ts compiler returns emitSkipped', async () => {
+  const { err, stdout } = await exec(
+    `${CMD_TS_NODE_WITHOUT_PROJECT_FLAG} ./outside-rootDir/foo.js`,
+    {
+      cwd: join(TEST_DIR, 'emit-skipped-fallback'),
+    }
+  );
+  expect(err).toBe(null);
+  expect(stdout).toBe('foo\n');
+});

--- a/tests/emit-skipped-fallback/outside-rootDir/foo.js
+++ b/tests/emit-skipped-fallback/outside-rootDir/foo.js
@@ -1,0 +1,15 @@
+// This file causes TS to return emitSkipped because it's outside of rootDir and
+// it's .js.  I assume this happens because the emit path is the same as the
+// input path, and perhaps also because the file is classified "external"
+
+const decorator = () => {};
+
+class Foo {
+  // Using a decorator to prove this .js file is getting compiled
+  @decorator
+  method() {
+    return 'foo';
+  }
+}
+
+console.log(new Foo().method());

--- a/tests/emit-skipped-fallback/tsconfig.json
+++ b/tests/emit-skipped-fallback/tsconfig.json
@@ -1,0 +1,7 @@
+{
+  "compilerOptions": {
+    "allowJs": true,
+    "experimentalDecorators": true,
+    "rootDir": "rootDir"
+  }
+}

